client/src/controllers/SwapController.ts

```diff
@@ -0,0 +1,192 @@
+import { Controller } from '@hotwired/stimulus';
+import { debounce } from '../utils/debounce';
+
+declare global {
+  interface Window {
+    headerSearch?: { targetOutput: string; url: string };
+  }
+}
+
+/**
+ * Allow for an element to trigger an async query that will
+ * path the results into a results DOM container. The query
+ * input can be the controlled element or the containing form.
+ */
+export class SwapController extends Controller<
+  HTMLFormElement | HTMLInputElement
+> {
+  static targets = ['input'];
+
+  static values = {
+    icon: { default: '', type: String },
+    loading: { default: false, type: Boolean },
+    src: { default: '', type: String },
+    target: { default: '#results', type: String },
+    wait: { default: 200, type: Number },
+  };
+
+  declare iconValue: string;
+  declare loadingValue: boolean;
+  declare readonly hasInputTarget: boolean;
+  declare readonly hasTargetValue: boolean;
+  declare readonly hasUrlValue: boolean;
+  declare readonly inputTarget: HTMLInputElement;
+  declare srcValue: string;
+  declare targetValue: string;
+  declare waitValue: number;
+
+  /** Allow cancelling of in flight async request if disconnected */
+  abortController?: AbortController;
+  /** The related icon element to attach the spinner to */
+  iconElement?: SVGUseElement | null;
+  /** Element that receives the fetch result HTML output */
+  targetElement?: HTMLElement;
+  /** Debounced function to fetch results & patch into the DOM */
+  update?: { (...args: any[]): void; cancel(): void };
+
+  connect() {
+    const formContainer = this.hasInputTarget
+      ? this.inputTarget.form
+      : this.element;
+
+    // support legacy window global approach (add warning in the future)
+    const { url: src, targetOutput } = window.headerSearch || {};
+    this.targetValue = targetOutput || this.targetValue;
+    this.srcValue =
+      src || this.srcValue || formContainer?.getAttribute('action') || '';
+    this.iconElement = null;
+
+    const targeElement = document.querySelector(this.targetValue);
+
+    const foundTarget = targeElement && targeElement instanceof HTMLElement;
+    const hasValidUrlValue = !!this.srcValue;
+
+    if (!foundTarget || !hasValidUrlValue) {
+      throw new Error(
+        [
+          !hasValidUrlValue && 'Cannot find valid src URL value',
+          !foundTarget &&
+            `Cannot find valid target element at "${this.targetValue}"`,
+        ]
+          .filter(Boolean)
+          .join(', '),
+      );
+    }
+
+    this.targetElement = targeElement as HTMLElement;
+
+    // set up icons
+    const iconContainer = (
+      this.hasInputTarget ? this.inputTarget : this.element
+    ).parentElement;
+
+    this.iconElement = iconContainer?.querySelector('use') || null;
+    this.iconValue = this.iconElement?.getAttribute('href') || '';
+    this.loadingValue = false;
+
+    // set up debounced update method
+    this.update = debounce(this.swapLocation.bind(this), this.waitValue);
```

Comment:
This should be simpler. Just wrap the existing method without the rename. 

```suggestion
    this.swapLocation = debounce(this.swapLocation.bind(this), this.waitValue);
```

Then .. use that in the html. 

```
data-action="change->w-swap#swapLocation"
```